pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/LLCSegmentCommit.java

```diff
@@ -0,0 +1,152 @@
+    final SegmentCompletionManager segmentCompletionManager = getSegmentCompletionManager();
+
+    SegmentCompletionProtocol.Response response = segmentCompletionManager.segmentCommitStart(_segmentNameStr,
+        _instanceId, _offset);
+    if (response.equals(SegmentCompletionProtocol.RESP_COMMIT_CONTINUE)) {
+
+      // Get the segment and put it in the right place.
+      boolean success = uploadSegment(_instanceId, _segmentNameStr);
+
+      response = segmentCompletionManager.segmentCommitEnd(_segmentNameStr, _instanceId, _offset, success);
+    }
+
+    LOGGER.info("Response: instance={}  segment={} status={} offset={}", _instanceId, _segmentNameStr, response.getStatus(), response.getOffset());
+    return new StringRepresentation(response.toJsonString());
+  }
+
+  boolean extractParams() {
+    final String offsetStr = getReference().getQueryAsForm().getValues(SegmentCompletionProtocol.PARAM_OFFSET);
+    final String segmentName = getReference().getQueryAsForm().getValues(SegmentCompletionProtocol.PARAM_SEGMENT_NAME);
+    final String instanceId = getReference().getQueryAsForm().getValues(SegmentCompletionProtocol.PARAM_INSTANCE_ID);
+
+    if (offsetStr == null || segmentName == null || instanceId == null) {
+      LOGGER.error("Invalid call: offset={}, segmentName={}, instanceId={}", offsetStr, segmentName, instanceId);
+      return false;
+    }
+    _segmentNameStr = segmentName;
+    _instanceId = instanceId;
+    try {
+      _offset = Long.valueOf(offsetStr);
+    } catch (NumberFormatException e) {
+      LOGGER.error("Invalid offset {} for segment {} from instance {}", offsetStr, segmentName, instanceId);
+      return false;
+    }
+    return true;
+  }
+
+  boolean uploadSegment(final String instanceId, final String segmentNameStr) {
+    // 1/ Create a factory for disk-based file items
+    final DiskFileItemFactory factory = new DiskFileItemFactory();
+
+    // 2/ Create a new file upload handler based on the Restlet
+    // FileUpload extension that will parse Restlet requests and
+    // generates FileItems.
+    final RestletFileUpload upload = new RestletFileUpload(factory);
+    final List<FileItem> items;
+
+    try {
+      // The following statement blocks until the entire segment is read into memory.
+      items = upload.parseRequest(getRequest());
+
+      boolean found = false;
+      File dataFile = null;
+
+      for (final Iterator<FileItem> it = items.iterator(); it.hasNext() && !found; ) {
+        final FileItem fi = it.next();
+        if (fi.getFieldName() != null && fi.getFieldName().equals(segmentNameStr)) {
+          found = true;
+          dataFile = new File(tempDir, segmentNameStr);
+          fi.write(dataFile);
+        }
+      }
+
+      if (!found) {
+        LOGGER.error("Segment not included in request. Instance {}, segment {}", instanceId, segmentNameStr);
+        return false;
+      }
+      // TODO Untar and check quota here? Maybe even check table name with that in the segment?
```

Comment:
We will handle the quotas as we do with realtime now. Quotas not enforced when the segments get created. Instead, they are counted towards the quota and enforced when an offline segment is posted.